tx_test.go

```diff
@@ -602,6 +602,108 @@ func TestTx_CopyFile_Error_Normal(t *testing.T) {
 	}
 }
 
+// Ensure that a update that releases pages in ranges between read transactions
+// is properly applied and rolled back.
+func TestTx_releaseRange(t *testing.T) {
+	// Set initial mmap size well beyond the limit we will hit in this
+	// test, since we are testing with long running read transactions
+	// and will deadlock if db.grow is triggered.
+	db := MustOpenWithOption(&bolt.Options{InitialMmapSize: os.Getpagesize() * 100})
+	defer db.MustClose()
+
+	bucket := "bucket"
+
+	put := func(key, value string) {
+		if err := db.Update(func(tx *bolt.Tx) error {
+			b, err := tx.CreateBucketIfNotExists([]byte(bucket))
+			if err != nil {
+				t.Fatal(err)
+			}
+			return b.Put([]byte(key), []byte(value))
+		}); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	del := func(key string) {
+		if err := db.Update(func(tx *bolt.Tx) error {
+			b, err := tx.CreateBucketIfNotExists([]byte(bucket))
+			if err != nil {
+				t.Fatal(err)
+			}
+			return b.Delete([]byte(key))
+		}); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	getWithTxn := func(txn *bolt.Tx, key string) []byte {
+		return txn.Bucket([]byte(bucket)).Get([]byte(key))
+	}
+
+	openReadTxn := func() *bolt.Tx {
+		readTx, err := db.Begin(false)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return readTx
+	}
+
+	checkWithReadTxn := func(txn *bolt.Tx, key string, wantValue []byte) {
+		value := getWithTxn(txn, key)
+		if !bytes.Equal(value, wantValue) {
+			t.Errorf("Wanted value to be %s for key %s, but got %s", wantValue, key, string(value))
+		}
+	}
+
+	rollback := func(txn *bolt.Tx) {
+		if err := txn.Rollback(); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	put("k1", "v1")
```

Comment:
the test sequence is deterministic. we could randomize the test sequence to increase our confidence. but i think it is OK for now. better than nothing.

Reply:
Yeah, definitely better than nothing.  Could leave behind a TODO comment about that.

Reply:
Good questions. Yes, the goal of this test is fairly modest-- Guarantee that we have coverage of a `releaseRange` being applied to the txid ranges between long running transactions and that we check that the read transactions can read values from the txids they are pinned to after the release.

I'll flesh out the documentation of this test to make this more clear.

I'm working on a separate `quick` (randomized) test that might be better used for soak testing.  I need to compare it with the existing `simulation_test` and then will send out an issue summarizing how it would work for review, and if there is interest in a PR, I'll submit one.

Reply:
@jpbetz thanks! make sense to me.